src/v/pandaproxy/schema_registry/storage.h

```diff
@@ -654,60 +655,67 @@ struct consume_to_store {
 
     ss::future<ss::stop_iteration> operator()(model::record_batch b) {
         if (!b.header().attrs.is_control()) {
-            b.for_each_record(*this);
+            co_await model::for_each_record(b, [this](model::record& rec) {
+                return (*this)(std::move(rec));
+            });
         }
         co_return ss::stop_iteration::no;
     }
 
-    void operator()(model::record record) {
+    ss::future<> operator()(model::record record) {
         auto key = record.release_key();
         auto key_type_str = from_json_iobuf<topic_key_type_handler<>>(
           key.share(0, key.size_bytes()));
 
         auto key_type = from_string_view<topic_key_type>(key_type_str);
         if (!key_type.has_value()) {
-            vlog(plog.warn, "Ignoring keytype: {}", key_type_str);
-            return;
+            vlog(plog.error, "Ignoring keytype: {}", key_type_str);
+            co_return;
         }
 
         switch (*key_type) {
         case topic_key_type::noop:
-            return;
+            co_return;
         case topic_key_type::schema:
-            return apply(
+            co_return co_await apply(
               from_json_iobuf<schema_key_handler<>>(std::move(key)),
               from_json_iobuf<schema_value_handler<>>(record.release_value()));
         case topic_key_type::config:
-            return apply(
+            co_return co_await apply(
               from_json_iobuf<config_key_handler<>>(std::move(key)),
               from_json_iobuf<config_value_handler<>>(record.release_value()));
         }
     }
 
-    void apply(schema_key key, schema_value val) {
-        vassert(key.magic == 1, "Schema key magic is unknown");
-        vlog(
-          plog.debug,
-          "Inserting key: subject: {}, version: {} ",
-          key.sub(),
-          key.version);
-        auto res = _store.insert(val.sub, val.schema, val.type);
-        vassert(res.id == val.id, "Schema_id mismatch");
-        vassert(res.version == val.version, "Schema_version mismatch");
+    ss::future<> apply(schema_key key, schema_value val) {
```

Comment:
Wondering why this was futurized if theres no async work in the apply methods, just a co_return.

Reply:
> Wondering why this was futurized if theres no async work in the apply methods, just a co_return.

I struggled to get the non-futurized `for_each_record` to return a value.

Reply:
what about using normal exception handling?

Reply:
To support sharding, the `store` is becoming an `ss::peering_sharded_service<store>`, cross-core operations will require the store to be futurized, meaning this will have async operations in it.